clang-tools-extra/clang-tidy/misc/UnusedParametersCheck.cpp

```diff
@@ -26,6 +29,17 @@ bool isOverrideMethod(const FunctionDecl *Function) {
     return MD->size_overridden_methods() > 0 || MD->hasAttr<OverrideAttr>();
   return false;
 }
+
+bool hasAttrAfterParam(const SourceManager *SourceManager,
+                       const ParmVarDecl *Param) {
+  for (const auto *Attr : Param->attrs()) {
+    if (SourceManager->isBeforeInTranslationUnit(Param->getLocation(),
+                                                 Attr->getLocation())) {
+      return true;
+    }
+  }
+  return false;
+}
```

Comment:
It is unfortunate that for declaration attributes we do not have the information that attr appeared before or after the decl.
I would argue, for simplicity, we filter out all annotated params irrespective of the position of attribute.

Dealing with raw source locations can quickly become problematic. For example with expanded locations:
```cpp
#define ANNOTATE_BEFORE(x) [[clang::lifetimebound]] x
#define ANNOTATE_AFTER(x) x [[clang::lifetimebound]]

int* foo(ANNOTATE_BEFORE(int* x)) {}
int* bar(ANNOTATE_AFTER(int* x)) {}
```
https://godbolt.org/z/brYMKfcxf
It is possible to make it work but I don't see the point of this complexity when we can have a simpler filter without much fallout

Reply:
I've run this example, and with my CL clang-tidy comments out the argument in the `ANNOTATE_BEFORE` case, while leaving the `ANNOTATE_AFTER` case as-is. Is that incorrect? In the former case, the grammar makes the attribute apply to the (unnamed) parameter. I've also verified the clang built with #118501 doesn't complain on this.

But maybe I'm missing some other problem?..

Reply:
The behaviour seems correct then. Please add this test case in the tests.
`isBeforeInTranslationUnit` seems to be the correct function here.

Reply:
Done. Thanks for the great test!